### PR TITLE
Feature/acceptance tests

### PR DESCRIPTION
**Pull request by alexeyklyukin**
Abuse of the lettuce acceptance tests in order to check the real-world scenarios when running Patroni. At the moment we just check the basic replication and failover, but adding new tests is quite easy. The drawbacks are that it takes some time to run the tests, and there is no support for parallel execution

### CODE REVIEW

File: features/terrain.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        while self.patroni_is_running(pg_name):
+            if not kill:
+                self.processes[pg_name].terminate()
+            else:
+                self.processes[pg_name].kill()
+            time.sleep(1)
+        self.log.get('pg_name') and self.log[pg_name].close()
+        del self.processes[pg_name]
+
+    def make_patroni_test_config(self, pg_name, output_dir):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l60)

File: features/terrain.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+            self.config[pg_name] = self.make_patroni_test_config(pg_name, self.output_dir)
+
+            p = subprocess.Popen(['python', 'patroni.py', self.config[pg_name]],
+                                 stdout=self.log[pg_name], stderr=subprocess.STDOUT, cwd=cwd)
+            if not (p and p.pid and p.poll() is None):
+                assert False, "PostgreSQL {0} is not running after being started".format(pg_name)
+            self.processes[pg_name] = p
+        # wait while patroni is available for queries, but not more than 10 seconds.
+        for tick in range(self.availability_check_time_limit):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Unused variable 'tick'](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l94)

File: features/terrain.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        return pg_name in self.processes and self.processes[pg_name].pid and (self.processes[pg_name].poll() is None)
+
+    def stop_patroni(self, pg_name, kill=False):
+        while self.patroni_is_running(pg_name):
+            if not kill:
+                self.processes[pg_name].terminate()
+            else:
+                self.processes[pg_name].kill()
+            time.sleep(1)
+        self.log.get('pg_name') and self.log[pg_name].close()
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self.log.get('pg_name')) and (self.log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l57)

File: features/basic_replication.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        '''Table (\w+) is present on (\w+) after (\d+) seconds'''
+        for i in range(int(max_replication_delay)):
+            if world.pctl.query(pg_name, "SELECT 1 FROM {0}".format(table_name), fail_ok=True) is not None:
+                break
+            sleep(1)
+        else:
+            assert False,\
+                "Table {0} is not present on {1} after {2} seconds".format(table_name, pg_name, max_replication_delay)
+
+    def check_role(self, step, pg_name, pg_role, max_promotion_timeout):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l45)

File: features/basic_replication.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+class BasicReplicationSteps(object):
+
+    def __init__(self, environ):
+        self.env = environ
+
+    def start_patroni(self, step, pg_name):
+        '''I start (\w+)'''
+        return world.pctl.start_patroni(pg_name)
+
+    def stop_patroni(self, step, pg_name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l19)

File: features/basic_replication.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        '''I add the table (\w+) to (\w+)'''
+        # parse the configuration file and get the port
+        try:
+            world.pctl.query(pg_name, "CREATE TABLE {0}()".format(table_name))
+        except pg.Error as e:
+            assert False, "Error creating table {0} on {1}: {2}".format(table_name, pg_name, e)
+
+    def table_is_present_on(self, step, table_name, pg_name, max_replication_delay):
+        '''Table (\w+) is present on (\w+) after (\d+) seconds'''
+        for i in range(int(max_replication_delay)):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Unused variable 'i'](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l37)

File: features/basic_replication.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+PATRONI_CONFIG = '{}.yml'
+
+
+@steps
+class BasicReplicationSteps(object):
+
+    def __init__(self, environ):
+        self.env = environ
+
+    def start_patroni(self, step, pg_name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l15)

File: features/basic_replication.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        '''I start (\w+)'''
+        return world.pctl.start_patroni(pg_name)
+
+    def stop_patroni(self, step, pg_name):
+        '''I shut down (\w+)'''
+        return world.pctl.stop_patroni(pg_name)
+
+    def kill_patroni(self, step, pg_name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l23)

File: features/terrain.py (hunk 1 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l60)

File: features/basic_replication.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        '''I kill (\w+)'''
+        return world.pctl.stop_patroni(pg_name, kill=True)
+
+    def add_table(self, step, table_name, pg_name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l27)

File: features/terrain.py (hunk 2 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Unused variable 'tick'](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l94)

File: features/basic_replication.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        '''I add the table (\w+) to (\w+)'''
+        # parse the configuration file and get the port
+        try:
+            world.pctl.query(pg_name, "CREATE TABLE {0}()".format(table_name))
+        except pg.Error as e:
+            assert False, "Error creating table {0} on {1}: {2}".format(table_name, pg_name, e)
+
+    def table_is_present_on(self, step, table_name, pg_name, max_replication_delay):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l35)

File: features/terrain.py (hunk 3 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self.log.get('pg_name')) and (self.log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2110318102/issues/source?bid=3127594&fileBranchId=3144623#l57)

File: features/basic_replication.py (hunk 4 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l45)

File: features/basic_replication.py (hunk 5 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l19)

File: features/basic_replication.py (hunk 6 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Unused variable 'i'](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l37)

File: features/basic_replication.py (hunk 7 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l15)

File: features/basic_replication.py (hunk 8 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l23)

File: features/basic_replication.py (hunk 9 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l27)

File: features/basic_replication.py (hunk 10 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2110318101/issues/source?bid=3127594&fileBranchId=3144623#l35)

Comment (a1exsh) [inline features/basic_replication.py]: Flooder :-P

File: features/terrain.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+        feature_dir = os.path.join(pctl.patroni_path, "features", "output", feature_name.encode('utf-8').replace(' ', '_'))
+        if os.path.exists(feature_dir):
+            shutil.rmtree(feature_dir)
+        os.makedirs(feature_dir)
+        self._output_dir = feature_dir
+
+    def _is_running(self, pg_name):
+        return pg_name in self._processes and self._processes[pg_name].pid and (self._processes[pg_name].poll() is None)
+
+    def _make_patroni_test_config(self, pg_name, output_dir):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827817/issues/source?bid=3127594&fileBranchId=3144623#l112)

File: features/terrain.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+                "Patroni instance is not available for queries after {0} seconds".format(max_wait_limit)
+
+    def stop(self, pg_name, kill=False):
+        while self._is_running(pg_name):
+            if not kill:
+                self._processes[pg_name].terminate()
+            else:
+                self._processes[pg_name].kill()
+            time.sleep(1)
+        self._log.get('pg_name') and self._log[pg_name].close()
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self._log.get('pg_name')) and (self._log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2114827817/issues/source?bid=3127594&fileBranchId=3144623#l68)

File: features/terrain.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+            self._config[pg_name] = self._make_patroni_test_config(pg_name, self._output_dir)
+
+            p = subprocess.Popen(['python', 'patroni.py', self._config[pg_name]],
+                                 stdout=self._log[pg_name], stderr=subprocess.STDOUT, cwd=cwd)
+            if not (p and p.pid and p.poll() is None):
+                assert False, "PostgreSQL {0} is not running after being started".format(pg_name)
+            self._processes[pg_name] = p
+        # wait while patroni is available for queries, but not more than 10 seconds.
+        for tick in range(max_wait_limit):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Unused variable 'tick'](https://www.codacy.com/app/cyberdemn/patroni/file/2114827817/issues/source?bid=3127594&fileBranchId=3144623#l53)

File: features/basic_replication.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+class BasicReplicationSteps(object):
+
+    def __init__(self, environ):
+        self.env = environ
+
+    def start_patroni(self, step, name):
+        '''I start (\w+)'''
+        return world.pctl.start(name)
+
+    def stop_patroni(self, step, name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l17)

File: features/basic_replication.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        '''Table (\w+) is present on (\w+) after (\d+) seconds'''
+        for i in range(int(max_replication_delay)):
+            if world.pctl.query(pg_name, "SELECT 1 FROM {0}".format(table_name), fail_ok=True) is not None:
+                break
+            sleep(1)
+        else:
+            assert False,\
+                "Table {0} is not present on {1} after {2} seconds".format(table_name, pg_name, max_replication_delay)
+
+    def check_role(self, step, pg_name, pg_role, max_promotion_timeout):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l43)

File: features/basic_replication.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        '''I add the table (\w+) to (\w+)'''
+        # parse the configuration file and get the port
+        try:
+            world.pctl.query(pg_name, "CREATE TABLE {0}()".format(table_name))
+        except pg.Error as e:
+            assert False, "Error creating table {0} on {1}: {2}".format(table_name, pg_name, e)
+
+    def table_is_present_on(self, step, table_name, pg_name, max_replication_delay):
+        '''Table (\w+) is present on (\w+) after (\d+) seconds'''
+        for i in range(int(max_replication_delay)):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Unused variable 'i'](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l35)

File: features/basic_replication.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+    def add_table(self, step, table_name, pg_name):
+        '''I add the table (\w+) to (\w+)'''
+        # parse the configuration file and get the port
+        try:
+            world.pctl.query(pg_name, "CREATE TABLE {0}()".format(table_name))
+        except pg.Error as e:
+            assert False, "Error creating table {0} on {1}: {2}".format(table_name, pg_name, e)
+
+    def table_is_present_on(self, step, table_name, pg_name, max_replication_delay):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l33)

File: features/basic_replication.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+from lettuce import world, steps
+
+
+@steps
+class BasicReplicationSteps(object):
+
+    def __init__(self, environ):
+        self.env = environ
+
+    def start_patroni(self, step, name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l13)

File: features/basic_replication.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+        '''I start (\w+)'''
+        return world.pctl.start(name)
+
+    def stop_patroni(self, step, name):
+        '''I shut down (\w+)'''
+        return world.pctl.stop(name)
+
+    def kill_patroni(self, step, name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l21)

File: features/basic_replication.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+        '''I kill (\w+)'''
+        return world.pctl.stop(name, kill=True)
+
+    def add_table(self, step, table_name, pg_name):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827816/issues/source?bid=3127594&fileBranchId=3144623#l25)

File: features/terrain.py [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+                "Patroni instance is not available for queries after {0} seconds".format(max_wait_limit)
+
+    def stop(self, pg_name, kill=False):
+        while self._is_running(pg_name):
+            if not kill:
+                self._processes[pg_name].terminate()
+            else:
+                self._processes[pg_name].kill()
+            time.sleep(1)
+        self._log.get('pg_name') and self._log[pg_name].close()
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self._log.get('pg_name')) and (self._log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2115469249/issues/source?bid=3127594&fileBranchId=3144623#l68)

File: features/terrain.py [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+
+            self._config[pg_name] = self._make_patroni_test_config(pg_name)
+
+            p = subprocess.Popen(['python', 'patroni.py', self._config[pg_name]],
+                                 stdout=self._log[pg_name], stderr=subprocess.STDOUT, cwd=cwd)
+            if not (p and p.pid and p.poll() is None):
+                assert False, "PostgreSQL {0} is not running after being started".format(pg_name)
+            self._processes[pg_name] = p
+        # wait while patroni is available for queries, but not more than 10 seconds.
+        for tick in range(max_wait_limit):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Unused variable 'tick'](https://www.codacy.com/app/cyberdemn/patroni/file/2115469249/issues/source?bid=3127594&fileBranchId=3144623#l53)

File: features/basic_replication.py (hunk 14 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l17)

File: features/basic_replication.py (hunk 15 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l43)

File: features/basic_replication.py (hunk 16 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Unused variable 'i'](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l35)

File: features/basic_replication.py (hunk 17 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l33)

File: features/basic_replication.py (hunk 18 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l13)

File: features/basic_replication.py (hunk 19 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l21)

File: features/basic_replication.py (hunk 20 above)
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2115469248/issues/source?bid=3127594&fileBranchId=3144623#l25)

File: features/basic_replication.py [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        # parse the configuration file and get the port
+        try:
+            world.pctl.query(pg_name, "CREATE TABLE {0}()".format(table_name))
+        except pg.Error as e:
+            assert False, "Error creating table {0} on {1}: {2}".format(table_name, pg_name, e)
+
+    @staticmethod
+    def table_is_present_on(step, table_name, pg_name, max_replication_delay):
+        '''Table (\w+) is present on (\w+) after (\d+) seconds'''
+        for i in range(int(max_replication_delay)):
```
Comment (CyberDem0n) [inline features/basic_replication.py]: Codacy Issue found: [Unused variable 'i'](https://www.codacy.com/app/cyberdemn/patroni/file/2115598838/issues/source?bid=3127594&fileBranchId=3144623#l40)

File: features/terrain.py (hunk 21 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self._log.get('pg_name')) and (self._log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2115598839/issues/source?bid=3127594&fileBranchId=3144623#l68)

File: features/terrain.py (hunk 22 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Unused variable 'tick'](https://www.codacy.com/app/cyberdemn/patroni/file/2115598839/issues/source?bid=3127594&fileBranchId=3144623#l53)

File: features/terrain.py (hunk 21 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self._log.get('pg_name')) and (self._log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2115708395/issues/source?bid=3127594&fileBranchId=3144623#l68)

File: features/terrain.py (hunk 22 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Unused variable 'tick'](https://www.codacy.com/app/cyberdemn/patroni/file/2115708395/issues/source?bid=3127594&fileBranchId=3144623#l53)

File: features/terrain.py [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+                "Patroni instance is not available for queries after {0} seconds".format(max_wait_limit)
+
+    def stop(self, pg_name, kill=False):
+        while self._is_running(pg_name):
+            if not kill:
+                self._processes[pg_name].terminate()
+            else:
+                self._processes[pg_name].kill()
+            time.sleep(1)
+        self._log.get('pg_name') and self._log[pg_name].close()
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Expression "(self._log.get('pg_name')) and (self._log[pg_name].close())" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2115736694/issues/source?bid=3127594&fileBranchId=3144623#l68)

File: features/terrain.py [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+        self.handle =\
+            subprocess.Popen(["etcd", "--data-dir", self.work_directory], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        start_time = time.time()
+        while (not self._is_running()):
+            if time.time() - start_time > self.start_timeout:
+                assert False, "Failed to start etcd"
+            time.sleep(1)
+        return True
+
+    def query(self, key):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2125793123/issues/source?bid=3127594&fileBranchId=3144623#l193)

File: features/patroni_api.py [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+        '''I receive a response (\w+) (.*)'''
+        if component == 'code':
+            assert self.status_code == int(data), "status code {0} != {1}".format(self.status_code, int(data))
+        elif component == 'text':
+            assert self.response == data.strip('"'), "response {0} does not contain {1}".format(self.response, data)
+        else:
+            assert component in self.response, "{0} is not part of the response".format(component)
+            assert self.response[component] == data, "{0} does not contain {1}".format(component, data)
+
+    def replication_works(self, step, time_limit):
```
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2126016496/issues/source?bid=3127594&fileBranchId=3144623#l82)

File: features/terrain.py (hunk 25 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2126016497/issues/source?bid=3127594&fileBranchId=3144623#l193)

File: features/terrain.py [hunk 27]
```diff
@@ -0,0 +1,285 @@
...
+        r = None
+        try:
+            r = requests.delete(EtcdController.ETCD_CLEANUP_URL)
+            if r and not r.ok:
+                assert False,\
+                    "request to cleanup the etcd contents was not successfull: status code {0}".format(r.status_code)
+        except requests.exceptions.RequestException as e:
+            assert False, "exception when cleaning up etcd contents: {0}".format(e)
+
+    def _is_running(self):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2152073347/issues/source?bid=3127594&fileBranchId=3144623#l241)

File: features/terrain.py [hunk 28]
```diff
@@ -0,0 +1,285 @@
...
+            subprocess.Popen(["etcd", "--debug", "--data-dir", self.work_directory],
+                             stdout=self.log_file, stderr=subprocess.STDOUT)
+        start_time = time.time()
+        while (not self._is_running()):
+            if time.time() - start_time > self.start_timeout:
+                assert False, "Failed to start etcd"
+            time.sleep(1)
+        return True
+
+    def query(self, key):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2152073347/issues/source?bid=3127594&fileBranchId=3144623#l201)

File: features/patroni_api.py [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
```
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2152073346/issues/source?bid=3127594&fileBranchId=3144623#l82)

File: features/patroni_api.py [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+            assert self.response[component] == data, "{0} does not contain {1}".format(component, data)
+
+    def replication_works(self, step, master, replica, time_limit):
+        '''replication works from (\w+) to (\w+) after (\d+) seconds'''
+        step.behave_as("""
+           When I add the table test_{0} to {1}
+            Then table test_{0} is present on {2} after {3} seconds
+            """.format(int(time.time()), master, replica, time_limit))
+
+    def scheduld_failover(self, step, at_url, from_host, to_host, in_seconds):
```
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2155987801/issues/source?bid=3127594&fileBranchId=3144623#l92)

File: features/patroni_api.py [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+        if component == 'code':
+            assert self.status_code == int(data),\
+                    "status code {0} != {1}, response: {2}".format(self.status_code, int(data), self.response)
+        elif component == 'text':
+            assert self.response == data.strip('"'), "response {0} does not contain {1}".format(self.response, data)
+        else:
+            assert component in self.response, "{0} is not part of the response".format(component)
+            assert self.response[component] == data, "{0} does not contain {1}".format(component, data)
+
+    def replication_works(self, step, master, replica, time_limit):
```
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2155987801/issues/source?bid=3127594&fileBranchId=3144623#l85)

File: features/terrain.py (hunk 27 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2155987802/issues/source?bid=3127594&fileBranchId=3144623#l241)

File: features/terrain.py (hunk 28 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2155987802/issues/source?bid=3127594&fileBranchId=3144623#l201)

File: features/patroni_api.py (hunk 30 above)
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2212627071/issues/source?bid=3127594&fileBranchId=3144623#l92)

File: features/patroni_api.py (hunk 31 above)
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2212627071/issues/source?bid=3127594&fileBranchId=3144623#l85)

File: features/terrain.py [hunk 32]
```diff
@@ -0,0 +1,307 @@
...
+        r = None
+        try:
+            r = requests.delete(EtcdController.ETCD_CLEANUP_URL)
+            if r and not r.ok:
+                assert False,\
+                    "request to cleanup the etcd contents was not successfull: status code {0}".format(r.status_code)
+        except requests.exceptions.RequestException as e:
+            assert False, "exception when cleaning up etcd contents: {0}".format(e)
+
+    def _is_running(self):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2212627072/issues/source?bid=3127594&fileBranchId=3144623#l263)

File: features/terrain.py [hunk 33]
```diff
@@ -0,0 +1,307 @@
...
+            subprocess.Popen(["etcd", "--debug", "--data-dir", self.work_directory],
+                             stdout=self.log_file, stderr=subprocess.STDOUT)
+        start_time = time.time()
+        while (not self._is_running()):
+            if time.time() - start_time > self.start_timeout:
+                assert False, "Failed to start etcd"
+            time.sleep(1)
+        return True
+
+    def query(self, key):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2212627072/issues/source?bid=3127594&fileBranchId=3144623#l223)

File: features/terrain.py (hunk 32 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2215820088/issues/source?bid=3127594&fileBranchId=3144623#l263)

File: features/terrain.py (hunk 33 above)
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2215820088/issues/source?bid=3127594&fileBranchId=3144623#l223)

File: features/patroni_api.py (hunk 30 above)
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2215820087/issues/source?bid=3127594&fileBranchId=3144623#l92)

File: features/patroni_api.py (hunk 31 above)
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2215820087/issues/source?bid=3127594&fileBranchId=3144623#l85)

File: features/terrain.py [hunk 34]
```diff
@@ -0,0 +1,312 @@
...
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2219520422/issues/source?bid=3127594&fileBranchId=3144623#l228)

File: features/terrain.py [hunk 35]
```diff
@@ -0,0 +1,312 @@
...
+        r = None
+        try:
+            r = requests.delete(EtcdController.ETCD_CLEANUP_URL)
+            if r and not r.ok:
+                assert False,\
+                    "request to cleanup the etcd contents was not successfull: status code {0}".format(r.status_code)
+        except requests.exceptions.RequestException as e:
+            assert False, "exception when cleaning up etcd contents: {0}".format(e)
+
+    def _is_running(self):
```
Comment (CyberDem0n) [inline features/terrain.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2219520422/issues/source?bid=3127594&fileBranchId=3144623#l268)

File: features/patroni_api.py [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+        '''replication works from (\w+) to (\w+) after (\d+) seconds'''
+        step.behave_as("""
+           When I add the table test_{0} to {1}
+            Then table test_{0} is present on {2} after {3} seconds
+            """.format(int(time.time()), master, replica, time_limit))
+
+    def scheduld_failover(self, step, at_url, from_host, to_host, in_seconds):
```
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2219520421/issues/source?bid=3127594&fileBranchId=3144623#l92)

File: features/patroni_api.py [hunk 37]
```diff
@@ -0,0 +1,99 @@
...
+        if component == 'code':
+            assert self.status_code == int(data),\
+                "status code {0} != {1}, response: {2}".format(self.status_code, int(data), self.response)
+        elif component == 'text':
+            assert self.response == data.strip('"'), "response {0} does not contain {1}".format(self.response, data)
+        else:
+            assert component in self.response, "{0} is not part of the response".format(component)
+            assert self.response[component] == data, "{0} does not contain {1}".format(component, data)
+
+    def replication_works(self, step, master, replica, time_limit):
```
Comment (CyberDem0n) [inline features/patroni_api.py]: Codacy Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2219520421/issues/source?bid=3127594&fileBranchId=3144623#l85)

**Pull request merged**